Add settler tile adjustments to Civilization class

bmgxyz opened the pull request:
- Add `SettlerTileAdjustments` as a child of `Civilization`
  - Encodes the previously hard-coded values as defaults for public fields
  - ~~Uses lambda functions to take values at runtime for adjustments that take arguments; this allows future users to supply arbitrary logic~~
- Replace hard-coded bonus values in `SettlerLocationAI.cs` with ~~calls and~~ references to `SettlerTileAdjustments` via the supplied `Player`
  - Notably, `DistancePenalty` is now a negative value by default, which I think is more clear because now adjustments are always added to the tile score
- Make settler tile score a `float` instead of `int`, which gives more precision for multiplicative adjustments
- ~~Apply lots of lints from Rider~~
  - ~~Identifier case tweaks (this caused most of the changes in other files)~~
  - ~~Using ternary operators to make code a little smaller when it's still clear~~
  - ~~Some grammar problems (Rider is picky about these for some reason)~~
  - ~~Lint ignores in cases where it makes sense (to me)~~
  - ~~I'm willing to undo these if they're too aggressive or unwanted~~

Closes #92

bmgxyz committed:
Add settler tile adjustments to Civilization class

- Add `SettlerTileAdjustments` as a child of `Civilization`
  - Encodes the previously hard-coded values as defaults for public fields
  - Uses lambda functions to take values at runtime for adjustments that take
    arguments; this allows future users to supply arbitrary logic
- Replace hard-coded bonus values in `SettlerLocationAI.cs` with calls and
  references to `SettlerTileAdjustments` via the supplied `Player`
  - Notably, `DistancePenalty` is now a negative value by default, which I
    think is more clear because now adjustments are always added to the tile
    score
- Make settler tile score a `float` instead of `int`, which gives more
  precision for multiplicative adjustments
- Apply lots of lints from Rider
  - Identifier case tweaks (this caused most of the changes in other files)
  - Using ternary operators to make code a little smaller when it's still clear
  - Some grammar problems (Rider is picky about these for some reason)
  - Lint ignores in cases where it makes sense (to me)
  - I'm willing to undo these if they're too aggressive or unwanted

TomWerner reviewed (changes requested):
Thanks for the PR, and the detailed description!

Would it be possible to split these apart into a couple of PRs? In particular, I think a PR that only fixes linting issues would be ideal. The linting issues touch so many files that its hard to see the actually interesting changes, so reducing diff noise would make it much easier to review the PR confidently.

Also it looks like `C7/C7.sln.DotSettings` is a new file, was that intended to be included in this PR?

bmgxyz commented:
> Would it be possible to split these apart into a couple of PRs? In particular, I think a PR that only fixes linting issues would be ideal.

Makes sense to me. I'll pull out the lint changes into a new PR and then go through the (many) other ones on that separate branch. It will take some time but will make for a cleaner development experience because the UI will quiet down. Only the parts that are directly related to #92 will remain here.

> Also it looks like `C7/C7.sln.DotSettings` is a new file, was that intended to be included in this PR?

This is part of the linting changes. Rider applies [spelling and grammar linting](https://www.jetbrains.com/help/rider/Proofreading.html) in comments :roll_eyes: and some of the words that appear in comments aren't in its dictionary (e.g., "civilopedia"). It's possible to add the unrecognized words to a dictionary so Rider stops flagging them, either locally (untracked) or "team-shared" (tracked). The file you mentioned is where Rider stores this configuration. I *think* tracking this file is a good idea because it *should* prevent this kind of linter noise in all Rider installations for this project.

A better option would probably be to find a way to disable this kind of linting entirely, but that should be a discussion elsewhere. I hadn't used Rider before this project, so I'm sure there's a lot I don't know here.

----------

Do you have any thoughts about my approach related to #92? I'm still unfamiliar with the project internals and C# in general, so it wouldn't surprise me if it can be improved.

TomWerner commented:
> Do you have any thoughts about my approach related to https://github.com/C7-Game/Prototype/issues/92? I'm still unfamiliar with the project internals and C# in general, so it wouldn't surprise me if it can be improved.

If you're talking about the changes to `Civilization.cs` and `SettlerLocationAI.cs`, yeah, that seems good. Do you happen to know if the tests pass successfully? I'm curious as to how the new class of constants is handled with JSON serialization.

bmgxyz committed:
Revert linting changes

bmgxyz committed:
Merge branch 'Development' into 0092-settler-ai-weights

bmgxyz committed:
Fix a few issues that I created during the merge

bmgxyz committed:
Fix tests by altering serialization and updating test saves

stavrosfa commented on C7Engine/C7GameData/Civilization.cs:
```diff
@@ -49,6 +50,24 @@ public Civilization(string name) {
+
+			// constants
+			public float HillsBonus = 10;
+			public float LuxuryResourceBonus = 15;
```
Could you add the notes that are above the constants in the original SettlerLocationAI regarding weights as a TODO here so we won't miss/forget it in the future?

stavrosfa commented on C7Engine/AI/UnitAI/SettlerLocationAI.cs:
```diff
@@ -1,24 +1,18 @@
 using System.Linq;
 using C7GameData.AIData;
 using Serilog;
+// ReSharper disable ForeachCanBeConvertedToQueryUsingAnotherGetEnumerator
```
Do we need these comments here?

bmgxyz replied:
Nope, good catch. These are left over from when I was trying to fix a lot of the lints that Rider was complaining about.

stavrosfa reviewed (commented):
Just some small stuff before Tom has a chance to review. I would wait for his review too, so you don't have to make multiple commits in case he wants some changes too

bmgxyz committed:
Remove lingering linter config comments, tweak imports, and re-add a TODO

bmgxyz committed:
Copy over the intended comment, mark as TODO, and link #802 in original TODO elsewhere

bmgxyz committed:
Rename SettlerLocationAi back to SettlerLocationAI

Capitalized the trailing I. Also, vim automatically removed some
trailing whitespace on unrelated comments.

bmgxyz committed:
Add SettlerLocationTest.cs

bmgxyz commented:
Added some tests

stavrosfa commented on EngineTests/AI/UnitAI/SettlerLocationTest.cs:
```diff
@@ -0,0 +1,137 @@
+
+namespace EngineTests.AI.UnitAI {
+	public sealed class SettlerLocationAITests : MapBase {
+		private Player MakeTestPlayer(List<Tile> knownTiles) {
```
This method and SurroundTile() look like usefull methods that could be used in other places in the future, wouldn't it be better to move them in MapBase?

stavrosfa commented on EngineTests/AI/UnitAI/SettlerLocationTest.cs:
```diff
@@ -0,0 +1,137 @@
+		}
+
+		[Fact]
+		private void HillsOverPlains() {
```
I was very confused after seeing the hills and plains tiles on the same location.
I was so sure that if I did
`Assert.True(hills.overlayTerrainType.isHilly());`
It would fail because it looks like it gets overriden by the plains tile later. But it passes, and I realized the reason is that the map contains all of the 18 tiles that are created with each pair being on the same location, because it's a list of tiles, not a hashset of some sort.
It works, and it should, given the API allows it, it's just confusing.
Perhaps put the tiles equally away from the Player, without having duplicate locations (at least the start tiles)? What do you think?

bmgxyz replied:
I was wondering if that would trip anyone up. It's a weird consequence of the design that this is possible, and I wonder if in the future we might try to eliminate it. My intention with the overlapping tiles in these tests was to be sure that distance wasn't contributing to the settler's decision at all. But I agree that it's not especially readable. I've separated the test locations and put the settler at the midpoint between them in the relevant tests. In my debugging I also found that there was a redundant distance penalty being applied, so I removed it.

stavrosfa replied:
Yeah it did trip me up good haha
I guess in the game it doesn't matter because the map is generated once and that's that, I don't think tiles are getting added or removed at any point.

> But I agree that it's not especially readable

Maybe a comment would have been enough for me not to go looking for ghosts in the code, although since I didn't know it ws permitted, I probably would have anyway

>there was a redundant distance penalty being applied, so I removed it

I don't know what's the logic behind it tbh, Tom is probably gonna have a look at it when he does his review

Everything else looks good to me :)

stavrosfa reviewed (commented):
Thanks for this!

bmgxyz committed:
Move MakeTestPlayer and SurroundTile into MapBase

bmgxyz committed:
Move SettlerLocationTest locations so they aren't on top of each other

bmgxyz committed:
Remove redundant distance penalty in SettlerLocationAI

TomWerner reviewed (commented):
Sorry for the delay, seems reasonable to me but I didn't have time to look in super close detail.

Seems like [stavrosfa](https://github.com/stavrosfa) has been doing a nicely thorough job - if they're happy with it, it sgtm
